[WMS][12.0] Add stock_putaway_abc - alpha version

grindtildeath opened the pull request:
This module implements ABC chaotic storage putaways.

Instead of defining putaway with a fixed location for product or product
category, this module allows to define an ABC preference for product or product
category on the product putaway form.

When such a putaway is applied, it will look recursively for a stock location
without children whose ABC classification matches the ABC priority of the
putaway.

e.g. for a putaway with B priority, it will look on children locations
for a B location. If no B location is found, it will look for a C. If no C
location is found, it will look for an A. Finally, if no location matches, it
will return the stock location where the putaway is applied.

Related to #691

grindtildeath committed:
12.0 create module stock putaway abc

grindtildeath committed:
fixup! 12.0 create module stock putaway abc

grindtildeath committed:
fixup! fixup! 12.0 create module stock putaway abc

jgrandguillaume commented:
Hi @grindtildeath 

Thanks for the work. I think either I've not understood something or we didn't understood each other.

What I want to do..

Location structure:

- Stock
  - HRL
      - Bin 1 Class A
      - Bin 2 Class B
      - Bin 3 Class C
  - FBG
      - Bin 1 Class A
      - Bin 2 Class B
      - Bin 3 Class C

Product put away:

- On Stock: Put away 1: Product A goes to HRL (set as a fixed location put away)
- On HRL: Put away 2: Product A is manage as a classification "A" (set as a ABC put away)
- On FBG: Put away 3: Product A is manage as a classification "C" (set as a ABC put away)

Expected result:

- If I receive product A in Stock (e.g. from a PO), it should process the put away and lead me to HRL/Bin 1 because put away 1 goes to HRL, then there the put away says apply ABC, so bin 1 is the match here
- If I receive product A in FBG (e.g. internal move), it should process the put away and lead me to FBG/Bin 3 because put away 3 says apply ABC, so bin 3 is the match here

I'm not sure how to achieve this with the current module. Should I merge put away 1 and 2 at stock level and create put away 3 for FBG ? How can I ask in the put away 1 that product should go to HRL then ?

Thanks for the headline, may be I'm just tired ;)

grindtildeath committed:
Depend on stock_putaway_rule

grindtildeath committed:
Merge abc putaway into stock.putaway.rule

grindtildeath committed:
[DROPME] oca_dependencies.txt

grindtildeath committed:
fixup! Merge abc putaway into stock.putaway.rule

Commit:
Fix method name

Commit:
Fix typos

grindtildeath committed:
fixup! fixup! Merge abc putaway into stock.putaway.rule

grindtildeath committed:
Fix error of extra call to find_abc_location

It is not needed since find_abc_location is already called in _get_putaway_strategy_abc

grindtildeath committed:
Loop on ABC putaway rules instead of getting the first

grindtildeath committed:
Override context key to make sure abc putaways are not filtered out

grindtildeath committed:
Improve constraint to avoid confusing putaways between fixed and abc
